File: tools/verify_transparency_batch_inputs_jws_vector.mjs
#!/usr/bin/env node

import { createPublicKey, verify } from "node:crypto";
import { readFileSync } from "node:fs";

const path = process.argv[2] ?? "test-vectors/transparency-batch-inputs-jws.json";
const vector = JSON.parse(readFileSync(path, "utf8"));
const fromBase64Url = (value) => Buffer.from(value, "base64url");

const canonicalize = (value) => {
  if (value === null || typeof value === "boolean" || typeof value === "string") return JSON.stringify(value);
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) throw new Error("Vector profile permits safe integers only");
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) return `[${value.map(canonicalize).join(",")}]`;
  if (typeof value === "object") {
    return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${canonicalize(value[key])}`).join(",")}}`;
  }
  throw new Error(`Unsupported JSON value: ${typeof value}`);
};

const protectedBytes = Buffer.from(canonicalize(vector.protected_header), "utf8");
const payloadBytes = Buffer.from(canonicalize(vector.payload), "utf8");
if (!protectedBytes.equals(fromBase64Url(vector.envelope.protected))) throw new Error("Protected JCS mismatch");
if (!payloadBytes.equals(fromBase64Url(vector.envelope.payload))) throw new Error("Payload JCS mismatch");
if (vector.protected_header.typ !== "application/xcim-transparency-batch-inputs+jcs") throw new Error("JWS typ mismatch");
if (vector.media_type !== "application/xcim-transparency-batch-inputs+jws+json") throw new Error("Media type mismatch");
if (vector.payload.object_type !== "transparency_batch_inputs" || vector.payload.batch_sequence !== 7 ||
    vector.payload.issuer_id !== "urn:xcim:issuer:emabled.com") throw new Error("Payload binding mismatch");

const signingInput = Buffer.from(`${vector.envelope.protected}.${vector.envelope.payload}`, "ascii");
if (signingInput.toString("ascii") !== vector.signing_input_ascii) throw new Error("Signing input mismatch");
const publicKey = createPublicKey({
  key: { kty: "OKP", crv: "Ed25519", x: vector.public_key_raw_base64url },
  format: "jwk",
});
if (!verify(null, signingInput, publicKey, fromBase64Url(vector.envelope.signature))) {
  throw new Error("Ed25519 signature verification failed");
}

const tamperedPayload = { ...vector.payload, batch_sequence: 8 };
const tamperedInput = Buffer.from(`${vector.envelope.protected}.${Buffer.from(canonicalize(tamperedPayload)).toString("base64url")}`, "ascii");
if (verify(null, tamperedInput, publicKey, fromBase64Url(vector.envelope.signature))) {
  throw new Error("Payload mutation preserved signature validity");
}

console.log(`Verified ${vector.vector_id}: JWS, JCS and sequence mutation rejection`);

File: tools/verify_transparency_batch_inputs_jws_vector.py
#!/usr/bin/env python3
"""Verify the candidate signed transparency-input vector independently."""

from __future__ import annotations

import base64
import json
from pathlib import Path
from typing import Any

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


ROOT = Path(__file__).resolve().parents[1]
VECTOR = ROOT / "test-vectors" / "transparency-batch-inputs-jws.json"
JWS_TYPE = "application/xcim-transparency-batch-inputs+jcs"


def decode(value: str) -> bytes:
    if not isinstance(value, str) or not value or "=" in value or any(
        char not in "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_" for char in value
    ):
        raise AssertionError("invalid unpadded base64url")
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


def canonicalize(value: Any) -> bytes:
    return json.dumps(value, ensure_ascii=False, allow_nan=False, separators=(",", ":"), sort_keys=True).encode("utf-8")


vector = json.loads(VECTOR.read_text(encoding="utf-8"))
envelope = vector["envelope"]
header_bytes = decode(envelope["protected"])
payload_bytes = decode(envelope["payload"])
header = json.loads(header_bytes)
payload = json.loads(payload_bytes)
assert header == vector["protected_header"] and canonicalize(header) == header_bytes
assert payload == vector["payload"] and canonicalize(payload) == payload_bytes
assert header["typ"] == JWS_TYPE and header["alg"] == "EdDSA"
public_key = decode(vector["public_key_raw_base64url"])
signature = decode(envelope["signature"])
assert len(public_key) == 32 and len(signature) == 64
Ed25519PublicKey.from_public_bytes(public_key).verify(
    signature, (envelope["protected"] + "." + envelope["payload"]).encode("ascii")
)
assert payload["object_type"] == "transparency_batch_inputs"
assert payload["issuer_id"] == "urn:xcim:issuer:emabled.com"
assert payload["batch_sequence"] == 7
print("Verified signed transparency-input vector independently in Python.")
